### ISSUE TITLE
Discuss Student Model

### ISSUE THREAD

**Issue by hassanhabib** [hassanhabib/OtripleS#1]
We need to conceptualize what a foundational student model properties are supposed to have.
Currently I'm thinking of the following:
```cs
    public class Student
    {
        public Guid Id { get; set; }
        public string FirstName { get; set; }
        public string MiddleName { get; set; }
        public string LastName { get; set; }
        public DateTimeOffset DateOfBirth { get; set; }
        public DateTimeOffset CreatedDate { get; set; }
        public Guid CreatedBy { get; set; }
        public DateTimeOffset UpdatedDate { get; set; }
        public Guid UpdatedBy { get; set; }
    }
```

If you think we need to add more or remove some of the aforementioned properties let's discuss, let us know what you think.
The current model will be locked for development by the end of the week on 28th of June 2020

**Comment by devmanzur** [comment]
ContactNumber, GuradianName, ClassTeacher, CurrentClass, RoleNumber,

**Comment by tajwal** [comment]
FatherName,Fathers email and mobile and gender.

**Comment by HatemGamal** [comment]
- the student's age when joined, religion, nationality, preferred language
- kin information
- Whether Handicap or not
- Other Notes to specific any disabilities or others

**Comment by devmanzur** [comment]
> * the student's age when joined, religion, nationality, preferred language
> * kin information
> * Whether Handicap or not
> * Other Notes to specific any disabilities or others

I'd say the date of birth would be more applicable instead of age.

**Comment by hassanhabib** [comment]
> > * the student's age when joined, religion, nationality, preferred language
> > * kin information
> > * Whether Handicap or not
> > * Other Notes to specific any disabilities or others
> 
> I'd say the date of birth would be more applicable instead of age.

I agree the date of birth is more dynamic than a static value that will require updating as time passes.

**Comment by hassanhabib** [comment]
> FatherName,Fathers email and mobile and gender of student.

@tajwal I wonder if these values should be in a different related entity, not directly in the main student entity.

**Comment by HatemGamal** [comment]
- A flag to show if the student is active or not. Reasons for being not active: graduation, disciplinary action, death... etc.
- Make a relation to siblings students.

**Comment by tajwal** [comment]
> > FatherName,Fathers email and mobile and gender of student.
> 
> @tajwal I wonder if these values should be in a different related entity, not directly in the main student entity.

Mr @hassanhabib, agrees with you In case we wants to have more relatives of a student then we can have an entity which will have one to many relation with student and will store more information like relation(father, brother, mother,..), First name,Last name,mobile,email, type(primary/secondary).

**Comment by tajwal** [comment]
> > FatherName,Fathers email and mobile and gender of student.
> 
> @tajwal I wonder if these values should be in a different related entity, not directly in the main student entity.

Mr @hassanhabib, Agrees with you In case when we wants to have more relatives of a student then we can have an entity which will have one to many relation with student and will store more information like relation(father, brother, mother,..), First name,Last name,mobile,email, and may be more columns.

**Comment by MCCshreyas** [comment]
Better we separate the AuditFields from domain model like last 4 properties from the above student model. 
We can introduce a new mode namely `BaseEntity` and can have those fields. So that we can use those fields in other domain models as well like School or Teachers. 
```csharp
public class BaseEntity
{
  public DateTimeOffset CreatedDate { get; set; }
  public Guid CreatedBy { get; set; }
  public DateTimeOffset UpdatedDate { get; set; }
  public Guid UpdatedBy { get; set; }
}
```
And whenever we need to use this just inherit the domain model from this base class. Like below.

```csharp 
public class Teacher : BaseEntity
{
  public string Name {get;set;}
}
```

**Comment by Driedas** [comment]
> > FatherName,Fathers email and mobile and gender of student.
> 
> @tajwal I wonder if these values should be in a different related entity, not directly in the main student entity.

As a parent myself, I think that there should be a list of contact people to reach out to in case of issues. Not even just the parents themselves, possibly grandparents or other people (maybe not even necessarily relatives) to be notified as well. I see this more as a m:n contact entity with contact type/relation to student. M;N for cases like having multiple children attending the same school. Can be simplified to 1:N if the identity of the contacts is not important (e. g. parent of child A is the same person as parent of child B)

**Comment by ahmetsekmen** [comment]
"StudentLevel" (first class , second third ..)

BaseEntity could be like this also

public class BaseEntity<T>
{
  public virtual T Id { get; set; }
  public DateTimeOffset CreatedDate { get; set; }
  public Guid CreatedBy { get; set; }
  public DateTimeOffset UpdatedDate { get; set; }
  public Guid UpdatedBy { get; set; }
}

**Comment by MCCshreyas** [comment]
@ahmetsekmen better to make Id as 'Guid' only, rather than taking T as generic one.
Because all users will be having guid as id, I can confirm this by looking at CreatedBy field as it will store guid of user who created that object or record!
So better to have guid explicitly.

**Comment by ahmetsekmen** [comment]
@MCCshreyas For users(admin,student,teacher..), guid is ok but what if other type of entities which they need audit properties.
Maybe we will decide to choose integer Id for those entities.
just an idea.

**Comment by MCCshreyas** [comment]
@ahmetsekmen Can you please provide example of other entity that you are trying to refer? That will be helpful for me to understand!

**Comment by ahmetsekmen** [comment]
@MCCshreyas 
BaseEntity;
public class BaseEntity< T >
{
 public virtual T Id { get; set; }
 public DateTimeOffset CreatedDate { get; set; }
 public Guid CreatedBy { get; set; }
 public DateTimeOffset UpdatedDate { get; set; }
 public Guid UpdatedBy { get; set; }
}

Here is how we will use;
public class Lesson : BaseEntity< int >
{
	public string LessonName { get; set; }
	public string LessonHour { get; set; }
}

For student;
public class Student : BaseEntity< Guid >
{
	public string FirsName { get; set; }
	public string MiddleName { get; set; }
	public string LastName { get; set; }
	...
	..
}

**Comment by devmanzur** [comment]
> @MCCshreyas
> BaseEntity;
> public class BaseEntity< T >
> {
> public virtual T Id { get; set; }
> public DateTimeOffset CreatedDate { get; set; }
> public Guid CreatedBy { get; set; }
> public DateTimeOffset UpdatedDate { get; set; }
> public Guid UpdatedBy { get; set; }
> }
> 
> Here is how we will use;
> public class Lesson : BaseEntity< int >
> {
> public string LessonName { get; set; }
> public string LessonHour { get; set; }
> }
> 
> For student;
> public class Student : BaseEntity< Guid >
> {
> public string FirsName { get; set; }
> public string MiddleName { get; set; }
> public string LastName { get; set; }
> ...
> ..
> }

I guess we all agree that there should be a base entity with the given fields. Let's discuss what other properties do we need in the student entity itself.

**Comment by devmanzur** [comment]
I guess guardians should be a separate entity itself, with courses, fees etc

**Comment by devmanzur** [comment]
we may add Admission Date, BatchId/SessionId, StudentImageUrl

**Comment by MCCshreyas** [comment]
@ahmetsekmen First a fall I need to highlight a thing that we should not use any domain specific property as primary key in database tables. 
I can see your Lesson domain entity 
Let's say I have following records.
1. Reflection
2. Generics
3. Functions

But let's say in future if school decided to swap those lessons, now they want Functions as first lesson, how will you do it now? As you have lessonno as primary key, you can't satisfy the new school requirements.

That's what I want to highlight is that we should never ever consider domain properties as primary key for tables although it may be unique. So we should always have our extra property that can acts as primary key.
Because you never know when domain requirement changes! 😉

**Comment by hassanhabib** [comment]
Very good points from everyone - please note that we need to separate additional details entities from information that might be inseparable from the main entity, for instance, a student date of birth is inseparable information,  their student identification number, but the guardians, contact information, teachers, classes all of this are related entities not part of the main entity we are trying to conceptualize here.

Also about the audit fields, from my personal experience I highly recommend maintaining each model fields within the model itself, as @ahmetsekmen pointed out, I have run into situations where audit fields are different from one entity to the other - also keeping these fields in the same model file makes it easier to understand what the model entails without having to jump between two files to get the full picture about what a particular model represents - thoughts?

**Comment by hassanhabib** [comment]
@a-urel might have some input on what a student entity is supposed to look like, care to share @a-urel ?

**Comment by MCCshreyas** [comment]
@hassanhabib its just a one inheritance chain for Audit fields to domain.
If we include every single field of audit in every domain object without a common one, that code will tend to duplicate.
And also consider a situation where we are introducing a soft delete feature in Application where IsDelete represent the status of delete then we need to add this fields to every single domain object by going to its individual file, but if we have a common base audit model then we can include that field there and it will start working for all of then domain models provided that they are inherited from it. 

And also the data coming into this fields will be most likely from a common method most likely it will be `dbcontext.SaveChangesAsync()`, if by mistake if any one missed a any audit field to include it in domain, it will throw a error at runtime.

So it's better to have common model for Audit rather than copying every single field to every single domain entity. Avoiding duplications. And also it's a *single level inheritance*, not much.
I will prefer jumping between two files rather than duplication. 😉

**Comment by hassanhabib** [comment]
@MCCshreyas  would you entertain the idea of an interface of audit fields? this way consistency is enforced and duplication is only generated, you don't have to type the audit fields, something like:

```csharp
public interface IAuditable {
     public DateTimeOffset CreatedDate {get; set;}
     public DateTimeOffset UpdatedDate {get; set;}
}


public class Student: IAuditable {
      public DateTimeOffset CreatedDate {get; set;}
      public DateTimeOffset UpdatedDate {get; set;}
}
```

**Comment by MCCshreyas** [comment]
@hassanhabib yeah it's OK. But does that mean that we need to copy and paste IAuditable to every single domain entity!? 🤔🤔🙄

**Comment by a-urel** [comment]
> @a-urel might have some input on what a student entity is supposed to look like, care to share @a-urel ?

@hassanhabib Since we need to conceptualize 'foundational' student model, we should keep it simple. 
I think it's useful to assume property values of this class don't change by student's academic status and activities. 
So it should contain only personal and legal information about students.

**Comment by hassanhabib** [comment]
> @hassanhabib yeah it's OK. But does that mean that we need to copy and paste IAuditable to every single domain entity, I mean in its file!? 🤔🤔🙄

@MCCshreyas nope, you just implement it and that's it - you don't have to copy the interface it's written once and used everywhere

**Comment by MCCshreyas** [comment]
@hassanhabib again it's two file right? Domain class and IAuditable interface file!

**Comment by tajwal** [comment]
> @MCCshreyas would you entertain the idea of an interface of audit fields? this way consistency is enforced and duplication is only generated, yet visibility of any model's properties are all aggregated in one single file. you don't have to type the audit fields, something like:
> 
> ```cs
> public interface IAuditable {
>      public DateTimeOffset CreatedDate {get; set;}
>      public DateTimeOffset UpdatedDate {get; set;}
> }
> 
> 
> public class Student: IAuditable {
>       public DateTimeOffset CreatedDate {get; set;}
>       public DateTimeOffset UpdatedDate {get; set;}
> }
> ```

Mr. @hassanhabib  what about something like below

**public interface IBaseModel** 
    {
        DateTime CreatedDate { get; set; }
        DateTime UpdatedDate { get; set; }
        int CreatedBy { get; set; }
        int UpdatedBy { get; set; }
        bool IsDelete { get; set; }
    }

**public class BaseModel : IBaseModel**
    {
        [Key]
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public int Id { get; set; }

        [Column(TypeName = "datetime2")]
        public DateTime CreatedDate { get; set; } = DateTime.Now;

        [Column(TypeName = "datetime2")]
        public DateTime UpdatedDate { get; set; } = DateTime.Now;
        public int CreatedBy { get; set; }
        public int UpdatedBy { get; set; }

        [DefaultValue(false)]
        public bool IsDelete { get; set; }
    }
	
**Public Class Student:BaseModel**
{
.........
}

**Comment by a-urel** [comment]
We can decide base classes or auditing strategies later.

> Premature optimization is the root of all evil

**Comment by devmanzur** [comment]
`public class Student : IAuditable
    {
        public Guid Id { get; set; }
        public string FirstName { get; set; }
        public string MiddleName { get; set; }
        public string LastName { get; set; }
        public DateTimeOffset DateOfBirth { get; set; }
        public DateTimeOffset AdmissionDate { get; set; }
        public Session Session { get; set; }
        public string ImageUrl { get; set; }
    }`

**Comment by devmanzur** [comment]
```
now

public class Student : IAuditable { public Guid Id { get; set; } public string FirstName { get; set; } public string MiddleName { get; set; } public string LastName { get; set; } public DateTimeOffset DateOfBirth { get; set; } public DateTimeOffset AdmissionDate { get; set; } public Session Session { get; set; } public string ImageUrl { get; set; } }
```

**Comment by devmanzur** [comment]
```
public class Student : IAuditable { 
  public Guid Id { get; set; }
  public string FirstName { get; set; }
  public string MiddleName { get; set; } 
  public string LastName { get; set; }
  public DateTimeOffset DateOfBirth { get; set; }
  public DateTimeOffset AdmissionDate { get; set; }
  public Session Session { get; set; }
  public string ImageUrl { get; set; }
 }
```

**Comment by hassanhabib** [comment]
> ```
> public class Student : IAuditable { 
>   public Guid Id { get; set; }
>   public string FirstName { get; set; }
>   public string MiddleName { get; set; } 
>   public string LastName { get; set; }
>   public DateTimeOffset DateOfBirth { get; set; }
>   public DateTimeOffset AdmissionDate { get; set; }
>   public Session Session { get; set; }
>   public string ImageUrl { get; set; }
>  }
> ```

@devmanzur looks good, I think Admission date and images should be in external entities, for instance here the U.S they have mutliple yearbook photos, so it should be external.
I think the rest is fine, but what is a session, and shouldn't you need a `SessionId` to fill in that related entitiy?

**Comment by antoniolinhart** [comment]
Should this class contain contact information like `emailAddress` or `phoneNumber`? What are the thoughts on how communication to the students will work, will it only be in-app?

Also, are we only supporting English? If not, `nativeLanguage` could be used to display the site in their native language. This might be more of a separate setting than a field for this class

**Comment by Mohamad-ali-8** [comment]
If Id for a student is Guid, I would recommend to also add another Reference number, an incremental number that starts somewhere. It is very handy later on for UI, searching purposes, and more meaningful for reporting. I found this very useful in one of the applications that I worked on.

public long ReferenceNumber { get; set; }

We added for it a sequence for DB like this:
   modelBuilder.Entity<Student>()
            .Property(o => o.ReferenceNumber)
            .HasDefaultValueSql("NEXT VALUE FOR dbo.ReferenceNumberSequence");

            modelBuilder
                .HasSequence<int>("ReferenceNumberSequence", schema: "dbo")
                .StartsAt(2365147).IncrementsBy(1);

I know I shouldn't think about the DB yet, but I just wanted to make it clearer.

**Comment by devmanzur** [comment]
> > ```
> > public class Student : IAuditable { 
> >   public Guid Id { get; set; }
> >   public string FirstName { get; set; }
> >   public string MiddleName { get; set; } 
> >   public string LastName { get; set; }
> >   public DateTimeOffset DateOfBirth { get; set; }
> >   public DateTimeOffset AdmissionDate { get; set; }
> >   public Session Session { get; set; }
> >   public string ImageUrl { get; set; }
> >  }
> > ```
> 
> @devmanzur looks good, I think Admission date and images should be in external entities, for instance here the U.S they have mutliple yearbook photos, so it should be external.
> I think the rest is fine, but what is a session, and shouldn't you need a `SessionId` to fill in that related entitiy?

Sessions/Batch to indicate which batch the user belongs to for instance I was session 2016. Batch 87. These attributes help to identify which sessions I was part of. Yeah, I missed the SessionId field.

**Comment by hiraldesai** [comment]
Demographics information like gender, race, etc.

1) are we planning on storing it?
2) if we are, are they considered l as “separable” information to the primary entity?

**Comment by Driedas** [comment]
What use cases is this model supposed to support? It's difficult to come up with a model when it's not clear what it should be doing...
The basic properties are there, requirements will tell us what needs to be added and what relationships there will be.

**Comment by hassanhabib** [comment]
> Demographics information like gender, race, etc.
> 
> 1. are we planning on storing it?
> 2. if we are, are they considered as “separable” information to the primary entity?

@hiraldesai  this is an interesting topic, I think gender and race are inseparable, but the question here would be why would we need to have that information at all in a student entity? one might argue maybe for particular curriculum in some countries, particular activities or such, and that might be a legitimate argument but from a global standpoint if we are building a schooling system generic enough to handle basic schooling needs I think these information might be on the side of external information rather than something fundamental to the entity to itself, what do you think?

**Comment by hassanhabib** [comment]
> What use cases is this model supposed to support? It's difficult to come up with a model when it's not clear what it should be doing...
> The basic properties are there, requirements will tell us what needs to be added and what relationships there will be.

@Driedas that's true. the one requirement we have for this issue is to conceptualize the basic information we need to have in a root entity for students - something to attach all other additional information to, despite what the additional information maybe depending on the requirement.
In other words, the student model should be a basic model/entity that contains basic information about a student in school - things every student must have like names, date of birth ... etc.

**Comment by hassanhabib** [comment]
> > > ```
> > > public class Student : IAuditable { 
> > >   public Guid Id { get; set; }
> > >   public string FirstName { get; set; }
> > >   public string MiddleName { get; set; } 
> > >   public string LastName { get; set; }
> > >   public DateTimeOffset DateOfBirth { get; set; }
> > >   public DateTimeOffset AdmissionDate { get; set; }
> > >   public Session Session { get; set; }
> > >   public string ImageUrl { get; set; }
> > >  }
> > > ```
> > 
> > 
> > @devmanzur looks good, I think Admission date and images should be in external entities, for instance here the U.S they have mutliple yearbook photos, so it should be external.
> > I think the rest is fine, but what is a session, and shouldn't you need a `SessionId` to fill in that related entitiy?
> 
> @hassanhabib Sessions/Batch to indicate which batch the user belongs to for instance I was session 2016. Batch 87. These attributes help to identify which sessions I was part of. Yeah, I missed the SessionId field.

In that case the session and batch are outside of the scope of this model, we are not there yet, we might come back and say let's do these things, but for now, we just need a simple basic student model to contain foundational information that are inseparable from any student entity anywhere in the world.
You might argue that sessions and batches are inseparable, and I would argue that there are exchange students that don't necessarily abide by these rules and don't necessarily need that information, what do you think?

**Comment by hassanhabib** [comment]
> If Id for a student is Guid, I would recommend to also add another Reference number, an incremental number that starts somewhere. It is very handy later on for UI, searching purposes, and more meaningful for reporting. I found this very useful in one of the applications that I worked on.
> 
> `public long ReferenceNumber { get; set; }`
> 
> We added for it a sequence for DB like this:
> 
> ```
>    modelBuilder.Entity<Student>()
>             .Property(o => o.ReferenceNumber)
>             .HasDefaultValueSql("NEXT VALUE FOR dbo.ReferenceNumberSequence");
> ```
> 
> ```
>         modelBuilder
>             .HasSequence<int>("ReferenceNumberSequence", schema: "dbo")
>             .StartsAt(2365147).IncrementsBy(1);
> ```
> 
> I know I shouldn't think about the DB yet, but I just wanted to make it clearer.

thanks @Mohamad-ali-8 - this is a good point - in that case why don't we just use a `ulong` as the data type for student ids instead of Guids? 

This is open for everyone to discuss.

**Comment by a-urel** [comment]
Actually student is a person, and we will have other person types like parents, teachers, personnel, etc. 
Maybe we will see that a base abstract Person class is needed in the future. But I think it's too early for this type normalizations.

**Comment by hassanhabib** [comment]
> Should this class contain contact information like `emailAddress` or `phoneNumber`? What are the thoughts on how communication to the students will work, will it only be in-app?
> 
> Also, are we only supporting English? If not, `nativeLanguage` could be used to display the site in their native language. This might be more of a separate setting than a field for this class
> 
> Edit: Saw contact info mentioned in other comment about related father contact info, but what about student contact info?
@antoniolinhart 
I think contact information are external entity information, simply because a given student might have multiple phone numbers for one student, maybe multiple e-mail addresses, contact information through other platforms, we definitely are going to need all of that, but I'm not sure if should be part of the main entity.

**Comment by hassanhabib** [comment]
It looks like the model that @devmanzur put together seems to be the closest to what we are going to go with, let me reiterate with a slight change:
```cs
public class Student : IAuditable { 
  public Guid Id { get; set; }
  public string FirstName { get; set; }
  public string MiddleName { get; set; } 
  public string LastName { get; set; }
  public DateTimeOffset DateOfBirth { get; set; }
  public DateTimeOffset CreatedDate {get; set;}
  public DateTimeOffset UpdatedDate {get; set;}
  public Guid CreatedBy {get; set;}
  public Guid UpdatedBy {get; set;}
 }
```

**Comment by Driedas** [comment]
> > Demographics information like gender, race, etc.
> > 
> > 1. are we planning on storing it?
> > 2. if we are, are they considered as “separable” information to the primary entity?
> 
> @hiraldesai this is an interesting topic, I think gender and race are inseparable, but the question here would be why would we need to have that information at all in a student entity? one might argue maybe for particular curriculum in some countries, particular activities or such, and that might be a legitimate argument but from a global standpoint if we are building a schooling system generic enough to handle basic schooling needs I think these information might be on the side of external information rather than something fundamental to the entity to itself, what do you think?

I don't see a use case for race, but I do see one for the sex of the student. This would come into play in coed environments when splitting up children into individual classes (not sure if e. g. the US system works like this, but in the European system that I'm mostly familiar with, kids in every year are split into groups of roughly equal sizes with roughly equal sex representation and stay in those groups for the majority of their time in the school), and for activities like physical education for reasons like gym/locker room capacities (PE is separated by sex, mostly due to locker rooms and different performance expectations, especially during and after puberty).

**Comment by Driedas** [comment]
@hassanhabib I am in favor of using GUIDs for the root aggregates as IDs
1) the ID can be chosen by the client making the student registration, there is no contention for database generated keys
2) makes synchronization between nodes in a distributed/not always online system easier (e. g. registration of students on a tablet with no internet connection, when done, the set of new students can easily be merged into the existing set)
3) the ID is usually a key in query strings when making queries, and this helps with preventing URL hacking to guess individual student IDs (of course this should not the only security measure, but is a nice addition essentially for free)

Enums and similar can be represented by integers.

**Comment by hassanhabib** [comment]
> @hassanhabib I am in favor of using GUIDs for the root aggregates as IDs
> 
> 1. the ID can be chosen by the client making the student registration, there is no contention for database generated keys
> 2. makes synchronization between nodes in a distributed/not always online system easier (e. g. registration of students on a tablet with no internet connection, when done, the set of new students can easily be merged into the existing set)
> 3. the ID is usually a key in query strings when making queries, and this helps with preventing URL hacking to guess individual student IDs (of course this should not the only security measure, but is a nice addition essentially for free)
> 
> Enums and similar can be represented by integers.

@Driedas very strong points - totally agree. anyone else has a counter argument?

**Comment by a-urel** [comment]
> It looks like the model that @devmanzur put together seems to be the closest to what we are going to go with, let me reiterate with a slight change:
> 
> ```cs
> public class Student : IAuditable { 
>   public Guid Id { get; set; }
>   public string FirstName { get; set; }
>   public string MiddleName { get; set; } 
>   public string LastName { get; set; }
>   public DateTimeOffset DateOfBirth { get; set; }
>   public DateTimeOffset CreatedDate {get; set;}
>   public DateTimeOffset UpdatedDate {get; set;}
>   public Guid CreatedBy {get; set;}
>   public Guid UpdatedBy {get; set;}
>  }
> ```
> 
> The one issue we want to discuss now, is what should the data type be for an Id? with reference to what @Mohamad-ali-8 talked about earlier.

Id field type of IdentityUser entity is string (nvarchar(450)). 
BTW, student is also a user and must be either derived from IdentityUser or linked to an IdentityUser.

**Comment by hassanhabib** [comment]
> > It looks like the model that @devmanzur put together seems to be the closest to what we are going to go with, let me reiterate with a slight change:
> > ```cs
> > public class Student : IAuditable { 
> >   public Guid Id { get; set; }
> >   public string FirstName { get; set; }
> >   public string MiddleName { get; set; } 
> >   public string LastName { get; set; }
> >   public DateTimeOffset DateOfBirth { get; set; }
> >   public DateTimeOffset CreatedDate {get; set;}
> >   public DateTimeOffset UpdatedDate {get; set;}
> >   public Guid CreatedBy {get; set;}
> >   public Guid UpdatedBy {get; set;}
> >  }
> > ```
> > 
> > 
> > The one issue we want to discuss now, is what should the data type be for an Id? with reference to what @Mohamad-ali-8 talked about earlier.
> 
> Id field type of IdentityUser entity is string (nvarchar(450)).
> BTW, student is also a user and must be either derived from IdentityUser or linked to an IdentityUser.

@a-urel I like the referencing rather than deriving - decoupling the business entity from a login/auth entity might be the better option - what do you think?

**Comment by Driedas** [comment]
student/guardian login credentials should definitely be a separate thing, there are products that already have their own schema for Authentication & Authorization (e. g. Identity server, ASP.NET Identity - or whatever the current iteration of the thing is called)

**Comment by a-urel** [comment]
@hassanhabib Will OtripleS satisfy all requirements of all educational institutions, from kindergarten to collage and universities?

**Comment by hassanhabib** [comment]
> @hassanhabib Will OtripleS satisfy all requirements of all educational institutions, from kindergarten to collage and universities?

@a-urel I was thinking the system should handle these levels:
Elementary(Primary School)
Middle School or Junior High
High School or Senior High School(Secondary School)

Kindergarten and universities or colleges seem to have different systems for instance college doesn't require a guardian or a parent, and kindergarten doesn't really have exams in a lot of countries.

**Comment by a-urel** [comment]
I think something like this would be enough:
```csharp
    public class Student : IAuditable
    {
        public Guid Id { get; set; }

        public byte[] Picture { get; set; }

        public string FirstName { get; set; }
        public string MiddleName { get; set; }
        public string LastName { get; set; }

        public DateTimeOffset BirthDate { get; set; }
        public int Gender { get; set; }

        /// <summary>
        /// Nationality, etc other demographic properties can be tags
        /// string or something like ICollection<Tag>...
        /// </summary>
        public string Tags { get; set; }

        public DateTimeOffset CreatedOn { get; set; }
        public string CreatedBy { get; set; }
        public DateTimeOffset ModifiedOn { get; set; }
        public string ModifiedBy { get; set; }
    }
```

**Comment by tobidub** [comment]
What about a Concurrency Token like timestamp or rowversion?

**Comment by levirgon** [comment]
> What about a Concurrency Token like timestamp or rowversion?

that might be a good candidate for another interface

**Comment by levirgon** [comment]
> I think something like this would be enough:
> 
> ```cs
>     public class Student : IAuditable
>     {
>         public Guid Id { get; set; }
> 
>         public byte[] Picture { get; set; }
> 
>         public string FirstName { get; set; }
>         public string MiddleName { get; set; }
>         public string LastName { get; set; }
> 
>         public DateTimeOffset BirthDate { get; set; }
>         public int Gender { get; set; } // or enum
> 
>         /// <summary>
>         /// Nationality, etc other demographic properties can be tags
>         /// string or something like ICollection<Tag>...
>         /// </summary>
>         public string Tags { get; set; }
> 
>         public DateTimeOffset CreatedOn { get; set; }
>         public string CreatedBy { get; set; } // nvarchar(450)
>         public DateTimeOffset ModifiedOn { get; set; }
>         public string ModifiedBy { get; set; }
>     }
> ```
@a-urel created on, modified on fields are. already inside IAuditable

**Comment by Driedas** [comment]
only a nitpick, I would change

>         public byte[] Picture { get; set; }

to something like 

`
public Photo/Image/Picture Picture { get ;set ; ]
`

maybe premature, but blobs are typically stored and retrieved separately regardless of the underlying storage technology, and there will most likely be photographs/pictures used elsewhere in the domain (plus there will be additional medata)

**Comment by a-urel** [comment]
> only a nitpick, I would change
> 
> > ```
> >     public byte[] Picture { get; set; }
> > ```
> 
> to something like
> 
> ```cs
> public Photo/Image/Picture Picture { get; set; }
> ```
> 
> maybe premature, but blobs are typically stored and retrieved separately regardless of the underlying storage technology, and there will most likely be photographs/pictures used elsewhere in the domain (plus there will be additional medata)
> 
> edit: formatting

Yes, Azure Blob Storage seems to be a better choice to store pictures and files.

**Comment by ahmetsekmen** [comment]
Keepin pics on azure server may can legally problem for schools.
I would say continue with byte.

**Comment by a-urel** [comment]
> > I think something like this would be enough:
> > ```cs
> >     public class Student : IAuditable
> >     {
> >         public Guid Id { get; set; }
> > 
> >         public byte[] Picture { get; set; }
> > 
> >         public string FirstName { get; set; }
> >         public string MiddleName { get; set; }
> >         public string LastName { get; set; }
> > 
> >         public DateTimeOffset BirthDate { get; set; }
> >         public int Gender { get; set; } // or enum
> > 
> >         /// <summary>
> >         /// Nationality, etc other demographic properties can be tags
> >         /// string or something like ICollection<Tag>...
> >         /// </summary>
> >         public string Tags { get; set; }
> > 
> >         public DateTimeOffset CreatedOn { get; set; }
> >         public string CreatedBy { get; set; } // nvarchar(450)
> >         public DateTimeOffset ModifiedOn { get; set; }
> >         public string ModifiedBy { get; set; }
> >     }
> > ```
> 
> @a-urel created on, modified on fields are. already inside IAuditable

or as @Driedas suggested:

```csharp
/// <summary>
/// Path or BlobName
/// Some DTOs may have Image or byte[] type fields
/// </summary>
public string Picture { get; set; }
```

**Comment by a-urel** [comment]
> Keepin pics on azure server may can be legally problem for schools.
> I would say let's continue with byte.

GDPR and other legal issues must be considered as well.

**Comment by Driedas** [comment]
> Keepin pics on azure server may can be legally problem for schools.
> I would say let's continue with byte.

this is not about whether to store those images in azure blobs, sql server varbinary(max) or on disk. Regardless of the underlying physical storage, there is typically metadata associated with an image (thus byte[] not sufficing), and creating an entity (in this case a student) and uploading binary data does not need to happen in the same transaction. The data for a student would be accessed more frequently and in different use cases than the associated image, hence why a separate entity makes sense to me. A string representing a URI to an image could also work (that URI can point to a table, blob, file, ...)

**Comment by Driedas** [comment]
> > Keepin pics on azure server may can be legally problem for schools.
> > I would say let's continue with byte.
> 
> GDPR and other legal issues must be considered as well.

GDPR etc is a good point...

**Comment by hidegh** [comment]
@a-urel  (not string Picture!) @Driedas nemtioned a class...

...and as there's an ORM (EF) planned to be used, the best way is always to keep picture **metadata** separately from the byte[] data, so the **Picture class** is necessary. Inside the meta you can have a link - which then can point to any kind of storage, you just implement a **strategy pattern** and IPictureDownloader + IPictureUpdater implementations.

Worth to consider if picture could be handled just as a **specific type of document** (and/or service), and due to GDPR store also a confidentiality level on them.

**Comment by AgentEnder** [comment]
Definitely going to be plenty of Metadata on images, and they will likely be used elsewhere (class cover photos and the like). They would definitely benefit from being another entity

**Comment by a-urel** [comment]
> @a-urel (not string Picture!) @Driedas nemtioned a class...
> 
> ...and as there's an ORM (EF) planned to be used, the best way is always to keep picture **metadata** separately from the byte[] data, so the **Picture class** is necessary. Inside the meta you can have a link - which then can point to any kind of storage, you just implement a **strategy pattern** and IPictureDownloader + IPictureUpdater implementations.
> 
> Worth to consider if picture could be handled just as a **specific type of document** (and/or service), and due to GDPR store also a confidentiality level on them - so this is why pure string is probably not the ideal solution (extensibility issues).

We may need a document / picture Entity (with metadata), and even versioning, etc. document management system features.
But I think profile pictures can be stored simply as string or byte[]. A profile picture entity could provide additional features like Gravatar though.

**Comment by Driedas** [comment]
byte[] properties on entities are problematic due to the necessity of loading the entire array into memory at once. For larger images (above ~85k, which is not too much considering the average resolution of phones/cameras) this would lead to those being stored on the large object heap, which a) the garbage collector is not too fond of collecting (only with Gen 2), b) are not compacted by the garbage collector thus causing memory fragmentation and requiring the runtime to request more and more memory from the OS. This is why storing something like a URI with the Picture class is preferable, to be able to stream the images in chunks that fit on the "regular" heap.

**Comment by a-urel** [comment]
I checked out some GitHub repos, including eShopWeb and some other e-commerce projects. I noticed two things:
* `string ImageUri` is preferred
* A backgroud job may also be needed to resize images to use as avatars or thumbnails

**Comment by devmanzur** [comment]
> > @hassanhabib I am in favor of using GUIDs for the root aggregates as IDs
> > 
> > 1. the ID can be chosen by the client making the student registration, there is no contention for database generated keys
> > 2. makes synchronization between nodes in a distributed/not always online system easier (e. g. registration of students on a tablet with no internet connection, when done, the set of new students can easily be merged into the existing set)
> > 3. the ID is usually a key in query strings when making queries, and this helps with preventing URL hacking to guess individual student IDs (of course this should not the only security measure, but is a nice addition essentially for free)
> > 
> > Enums and similar can be represented by integers.
> 
> @Driedas very strong points - totally agree. anyone else has a counter argument?

I am against database generated Ids as well. And the points by @Driedas are legit. So I think we should go with Guids as Ids.

**Comment by AgentEnder** [comment]
I prefer guids as well

**Comment by Akinnagbe** [comment]
> student/guardian/teacher login credentials should definitely be a separate thing, there are products that already have their own schema for Authentication & Authorization (e. g. Identity server, ASP.NET Identity - or whatever the current iteration of the thing is called)
> 
> edit: added teacher, cannot forget about the teachers :-)

using an IdentityServer could be a better option, in case we are considering mobile apps

**Comment by hassanhabib** [comment]
> > Keepin pics on azure server may can be legally problem for schools.
> > I would say let's continue with byte.
> 
> GDPR and other legal issues must be considered as well.

Definitely - we need to consider _**the right to be forgotten**_ as one of our priorities when it comes to data - everyone should have the right to have their data completely erased from any system - thank you for bringing this up, great point!

**Comment by hassanhabib** [comment]
> > student/guardian/teacher login credentials should definitely be a separate thing, there are products that already have their own schema for Authentication & Authorization (e. g. Identity server, ASP.NET Identity - or whatever the current iteration of the thing is called)
> > edit: added teacher, cannot forget about the teachers :-)
> 
> using an IdentityServer could be a better option, in case we are considering mobile apps

We definitely are considering mobile and desktop apps - we want to allow people to have offline mode where they can manage their local data - save notes and then sync when they get back online.

**Comment by a-urel** [comment]
My final draft:
```csharp
    public class Student : IAuditable
    {
        /// <summary>
        /// PK
        /// </summary>
        public Guid StudentId { get; set; }

        /// <summary>
        /// Ref. to related IdentityUser
        /// nvarchar(450)
        /// </summary>
        public string UserId { get; set; }

        /// <summary>
        /// Legal status of the student
        /// eg. citizen, guest student, etc.
        /// </summary>
        //public string LegalStatus { get; set; }

        /// <summary>
        /// National Identity Number
        /// </summary>
        public string IdentityNumber { get; set; }


        /// <summary>
        /// Path or BlobName
        /// Some DTOs may have Image or byte[] type fields
        /// </summary>
        public string Picture { get; set; }

        public string FirstName { get; set; }
        public string MiddleName { get; set; }
        public string LastName { get; set; }

        public DateTimeOffset BirthDate { get; set; }

        /// <summary>
        /// NotSpecified, Female, Male, Other
        /// </summary>
        public Gender Gender { get; set; }

        /// <summary>
        /// Nationality, etc other demographic properties can be tags
        /// string or something like ICollection<Tag>...
        /// </summary>
        //public string Tags { get; set; }

        //[Timestamp]
        //public byte[] RowVersion { get; set; }

        public DateTimeOffset CreatedOn { get; set; }
        public string CreatedBy { get; set; }
        public DateTimeOffset ModifiedOn { get; set; }
        public string ModifiedBy { get; set; }
    }
```

**Comment by hassanhabib** [comment]
Thank you everyone for your contributions, this is amazing - with the influence of each and every one of you here's the final model that we should start with:

I think this is good enough, just few modifications:
```csharp
public interface IAuditable {
	DateTimeOffset CreatedDate {get; set;}
	DateTimeOffset UpdatedDate {get; set;}
	Guid CreatedBy {get; set;}
	Guid UpdatedBy {get; set;}
}

public enum Gender {
	Female,
	Male,
	Other
}

public class Student : IAuditable
{
        public Guid Id { get; set; }
        public string UserId { get; set; }
        public string IdentityNumber { get; set; }
        public string FirstName { get; set; }
        public string MiddleName { get; set; }
        public string LastName { get; set; }
        public DateTimeOffset BirthDate { get; set; }
        public Gender Gender { get; set; }
        public DateTimeOffset CreatedDate { get; set; }
        public Guid CreatedBy { get; set; }
        public DateTimeOffset UpdatedDate { get; set; }
        public Guid UpdatedBy { get; set; }
}

```


Now, let's talk about assignments:

## Setup 

- Student Model Setup & Migrations with EF Core @a-urel  Monday 12:00 AM - 11:59 PM PST

## Broker

- Students Broker Select By Id function @AgentEnder by Tuesday 12:00 AM - 11:59 PM PST
- Students Broker Select All function @hiraldesai  by Tuesday 12:00 AM - 11:59 PM PST
- Students Broker Insert function @tajwal  by Tuesday 12:00 AM - 11:59 PM PST
- Students Broker Update function @hidegh by Tuesday 12:00 AM - 11:59 PM PST
- Students Broker Delete function @MChorfa by Tuesday 12:00 AM - 11:59 PM PST

## Service

- Student Service Register function @Driedas by Wednesday 12:00 AM - 11:59 PM PST
- Student Service Modify function @Akinnagbe by Wednesday 12:00 AM - 11:59 PM PST
- Student Service Retrieve by Id function @ahmetsekmen by Wednesday 12:00 AM - 11:59 PM PST
- Student Service Retrieve All function @HatemGamal by Wednesday 12:00 AM - 11:59 PM PST
- Student Service Delete by Id function @eriadhami by Wednesday 12:00 AM - 11:59 PM PST

## Controller

- Students Controller Post endpoint @MCCshreyas  by Thursday 12:00 AM - 11:59 PM PST
- Students Controller Get by Id endpoint @antoniolinhart by Thursday 12:00 AM - 11:59 PM PST
- Students Controller Get All endpoint @levirgon by Thursday 12:00 AM - 11:59 PM PST
- Students Controller Update endpoint @devmanzur by Thursday 12:00 AM - 11:59 PM PST
- Students Controller Delete endpoint @Mohamad-ali-8 by Thursday 12:00 AM - 11:59 PM PST


## Acceptance Tests

- Students Post Acceptance Test @tobidub by Friday 12:00 AM - 11:59 PM PST
- Students Update Acceptance Test @gauriramesh by Friday 12:00 AM - 11:59 PM PST
- Students Get All Acceptance Test @AliCharper by Friday 12:00 AM - 11:59 PM PST


Few Rules:
1. If you have a service task, make sure you commit a failing test, then commit making it pass and so on until your feature is complete.
2. all code must be PR'ed - do *not* push straight to master without the team looking at your PR first.
3. Take a look at how some of the things are done here: https://github.com/hassanhabib/SchoolEM and let me know if you have any questions about the implementation details.
4. This WIP Coding Standard https://github.com/hassanhabib/CSharpCodingStandard will be our guidance, it is a work in progress but it should give you a pretty good idea bout how our code should look like from guidelines and standardization standpoint.
5. Do not start your task before the designated time and make sure you finalize it at the expected time so others can build on top of your work.

If you haven't had an assignment this week please make sure you engage in discussion for next week so I know you are active in this repo and you are going to be able to contribute to the project.

As soon as all these tasks are done, this issue will be closed.

Thank you all from the bottom of my heart for your contributions, if you are stuck or unsure how to do a task, reach out to me literally on any social media platform you like, LinkedIn, Facebook, Whatsapp, Twitter or even Instagram or just ping me on gitter and I will be sure to response within 2 - 4 hours max.

**Comment by devmanzur** [comment]
@hassanhabib will we allow all properties to be updated?

**Comment by hassanhabib** [comment]
> @hassanhabib will we allow all properties to be updated?

@devmanzur yes, except of course the Id for obvious reasons

**Comment by devmanzur** [comment]
> > @hassanhabib will we allow all properties to be updated?
> 
> @devmanzur yes, except of course the Id for obvious reasons

Understood, Thank you.

**Comment by Mohamad-ali-8** [comment]
I will be working on the Delete Endpoint. Are we soft deleting or doing a physical delete?
Another not related question, are we going to use System versioned temporal tables?

```
using Microsoft.EntityFrameworkCore.Migrations;

namespace DHHS.RA.Data.Migrations
{
    public partial class MakeStudentTableTemporal : Migration
    {
        protected override void Up(MigrationBuilder migrationBuilder)
        {
            migrationBuilder.EnsureSchema("History");
            migrationBuilder.AddTemporalTableSupport("Student", "History");
        }

        protected override void Down(MigrationBuilder migrationBuilder)
        {
            migrationBuilder.EnsureSchema("History");
            migrationBuilder.TurnOffTemporalTableSupport("Student");
        }
    }
}
```

**Comment by Driedas** [comment]
I think that in very exceptional circumstances it makes sense to delete something, if ever. How about a status property to both student and teacher? Something like that will be useful anyway when students leave the school for any reason. Not sure what the values would be, but maybe something like Registered/Active/Graduated/Cancelled

**Comment by a-urel** [comment]
> I think that in only very exceptional circumstances it makes sense to delete something, if ever. How about a status property to both student and teacher? Something like that will be useful anyway when students leave the school for any reason. Not sure what the values would be, but maybe something like Registered/Active/Graduated/Cancelled

Exactly! States and transitions (state machine) are very useful in school information systems. Like:
Student: Candidate, Active, Graduated...
Semester: Draft, Active, Past...
Exam: Pending, Completed, Canceled...
...

**Comment by Driedas** [comment]
@a-urel cool, I like "Candidate" better than "Registered" :-)

**Comment by hassanhabib** [comment]
> I will be working on the Delete Endpoint. Are we soft deleting or doing a physical delete?
> Another not related question, are we going to use System versioned temporal tables?
> 
> ```
> using Microsoft.EntityFrameworkCore.Migrations;
> 
> namespace DHHS.RA.Data.Migrations
> {
>     public partial class MakeStudentTableTemporal : Migration
>     {
>         protected override void Up(MigrationBuilder migrationBuilder)
>         {
>             migrationBuilder.EnsureSchema("History");
>             migrationBuilder.AddTemporalTableSupport("Student", "History");
>         }
> 
>         protected override void Down(MigrationBuilder migrationBuilder)
>         {
>             migrationBuilder.EnsureSchema("History");
>             migrationBuilder.TurnOffTemporalTableSupport("Student");
>         }
>     }
> }
> ```

@Mohamad-ali-8  that's a good point - however, we want to expose a private endpoint to clean out the database from any records for acceptance testing - soft-deleting will be part of the processing services not these broker-neighboring services, the deleting here is actually for the system not for end-users to be able to delete records.
I think using temporal tables is a good idea, but this is not a part of the initial work we are doing here to get things started.

**Comment by hassanhabib** [comment]
> > I think that in only very exceptional circumstances it makes sense to delete something, if ever. How about a status property to both student and teacher? Something like that will be useful anyway when students leave the school for any reason. Not sure what the values would be, but maybe something like Registered/Active/Graduated/Cancelled
> 
> Exactly! States and transitions (state machine) are very useful in school information systems. Like:
> Student: Candidate, Active, Graduated...
> Semester: Draft (or Future), Active, Past...
> Exam: Pending, Completed, Canceled...
> ...

@a-urel That's a good idea, let's propose initial statuses for Students, here's a draft:
```csharp
public enum StudentStatus {
    Candidate,
    Enrolled,
    Graduated,
    Expelled,
    Deactivated
}

```

**Comment by a-urel** [comment]
Actually these states indicate the student's relationship with an academic program, which is usually the exact same thing with student's status. But a student may be registered to both CS and MBA, simultaneously or not. I'm not sure if it's too early to consider these details.

**Comment by hassanhabib** [comment]
> Actually these states indicate the student's relationship with an academic program, which is usually the exact same thing with student's status. But a student may be registered to both CS and MBA, simultaneously or not. I'm not sure if it's too early to consider these details.

@a-urel we are not doing college just schools

**Comment by a-urel** [comment]
> > Actually these states indicate the student's relationship with an academic program, which is usually the exact same thing with student's status. But a student may be registered to both CS and MBA, simultaneously or not. I'm not sure if it's too early to consider these details.
> 
> @a-urel we are not doing college just schools

In some countries a school may include both primary and secondary school (or secondary and high school) stages. 
To support this scenario we can either 
* assume these schools as separate organizations (and create a new student record for secondary school) or 
* provide many registrations (elementary, secondary,...) for single student.

First approach is simpler but the second one is more flexible.

These are the educational stages we should support:
https://en.wikipedia.org/wiki/Educational_stage

**Comment by a-urel** [comment]
@hassanhabib I think we can define **OtripleS** as **K-12** software:
https://en.wikipedia.org/wiki/K%E2%80%9312

Maybe another discussion issue would be useful, such as _scope_ or _big picture_.

**Comment by hidegh** [comment]
@a-urel exactly, **I do miss 2 important things on this project:** (was considering to create an ISSUE to address this, but...)
1. **the scope** - what it should include, features, goals, etc. - so the **big picture**
2. I personally find the talk about the model too early, **architecture and technologies used** should be the first, I know @hassanhabib pushed some sample code but:
- I personally do not favor the old MVC folder structure, as the **Feature folder** approach allows me to keep things belonging together beside each other.
- In the sample: https://github.com/hassanhabib/OtripleS/blob/master/OtripleS.Web.Api/Controllers/StudentsController.cs there's too much code in the controller, would consider **MediatR** with behaviors for logging, validation, exception handling, etc. - it's much cleaner design.

**Comment by hassanhabib** [comment]
@hidegh I created an issue for the architecture patterns and technology in this issue: https://github.com/hassanhabib/OtripleS/issues/33

Closing this issue as we currently have full CRUD operations as foundational functionality for Student model.
(We should add student status with a different issue later)
